firefox-ios/firefox-ios-tests/Tests/ClientTests/FoundationModelsSummarizerTests.swift

```diff
@@ -0,0 +1,132 @@
+            _ = try await subject.summarize(prompt: "p", text: "t")
+        }
+    }
+
+    func testSummarizerRespondStreaming() async throws {
+        let expectedResponse = ["a", "b", "c"]
+        let subject = createSubject(respondWith: expectedResponse)
+
+        var receivedChunks: [String] = []
+        var index = 0
+        let stream = subject.summarizeStreamed(prompt: "p", text: "t")
+        for try await chunk in  stream {
+            XCTAssertEqual(
+                chunk,
+                expectedResponse[index],
+                "chunk[\(index)] should be “\(expectedResponse[index])”"
+            )
+            receivedChunks.append(chunk)
+            index += 1
+        }
+
+        XCTAssertEqual(receivedChunks, expectedResponse)
+    }
+
+    func testSummarizerRespondStreamingThrowsGuardViolation() async throws {
+        let guardViolationError = LanguageModelSession.GenerationError.guardrailViolation(.init(debugDescription: "context"))
+        let subject = createSubject(respondWithError: guardViolationError)
+        let stream = subject.summarizeStreamed(prompt: "p", text: "t")
+
+        await assertSummarizeThrows(.safetyBlocked) {
+            // Consume the stream but do nothing
+            for try await _ in stream { }
+        }
+    }
+
+    func testSummarizerRespondStreamingThrowsUnknown() async throws {
+        let randomError = NSError(domain: "Random error", code: 1)
+        let subject = createSubject(respondWithError: randomError)
+        let stream = subject.summarizeStreamed(prompt: "p", text: "t")
+
+        await assertSummarizeThrows(.unknown(randomError)) {
+            // Consume the stream but do nothing
+            for try await _ in stream { }
+        }
+    }
+
+    private func createSubject(
+        respondWith responses: [String]? = nil,
+        respondWithError error: Error? = nil
+    ) -> FoundationModelsSummarizer {
+        let mockSession = MockLanguageModelSession()
+        if let responses {
+            mockSession.respondWith = responses
+        }
+        if let error {
+            mockSession.respondWithError = error
+        }
+        return FoundationModelsSummarizer(makeSession: { _ in mockSession })
+    }
+
+    /// Convenience method to simplify error checking in the test cases
+    private func assertSummarizeThrows(
+        _ expected: SummarizerError,
+        when running: @escaping () async throws -> Void,
+    ) async {
+        do {
+            try await running()
+            XCTFail("Expected summarize to throw, but it returned normally")
+        } catch let error as SummarizerError {
+            if error != expected {
+                XCTFail("Expected \(expected) to be thrown, but got \(error) instead")
+            }
+        } catch {
+            XCTFail("Expected SummarizerError, but got non SummarizerError: \(error)")
+        }
+    }
+}
+
+/// NOTE: Conforming Equatable to make checks in tests simpler.
```

Comment:
@cyndichin not super happy with this but it seems there is no simple way to check if errors are the same. Let me know what you think.

Reply:
Sweet thanks for the follow-up. I opened a ticket for [this](https://mozilla-hub.atlassian.net/browse/FXIOS-12971).